ext/LuxReactantExt/train.jl

```diff
@@ -0,0 +1,28 @@
+# Case III: Nothing is cached. First call to `single_train_step`
+function Lux.Experimental.single_train_step(
+        ad::AutoReactant, obj_fn::F, data, ts::Lux.Experimental.TrainState) where {F}
+    dps = Lux.recursive_make_zero(ts.parameters)
+
+    # Make reactant arrays
+    data = Lux.__make_reactant_array(data)
+    ps = Lux.__make_reactant_array(ts.parameters)
+    dps = Lux.__make_reactant_array(dps)
+    st = Lux.__make_reactant_array(ts.states)
+    st_opt = Lux.__make_reactant_array(ts.optimizer_state)
+
+    # obj_fn_wrapper, st_updated, stats = Lux.Experimental.__wrap_objective_function(
+    #     obj_fn, ts.model, ps, st, data, Val(true))
+
+    compiled_fn = Reactant.compile(
+        __update_fn_wrapper, (obj_fn, ts.model, ps, dps, st, st_opt, data))
+
+    # return
+end
+
+function __update_fn_wrapper(obj_fn, model, ps, dps, st, st_opt, data)
+    _, (loss, st_, stats) = Enzyme.autodiff(
```

Comment:
Minor, but this is going to be more efficient if dps is defined within the function and doesn't escape the scope.

Similarly it would be more efficient to have this do ps .= ps_ and not have to return the second allocation

Reply:
In other words all temporaries/allocations that cross the compiled function boundaries need to be materialized, but all variables defined only within can be optimized out

Reply:
Do you why I am able to return a `loss, st_, stats` here without enzyme complaining that I specified the return type as Active?

> In other words all temporaries/allocations that cross the compiled function boundaries need to be materialized, but all variables defined only within can be optimized out

Makes sense I will do that.